hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/InitMetaProcedure.java

```diff
@@ -71,7 +71,11 @@ private static void writeFsLayout(Path rootDir, Configuration conf) throws IOExc
     LOG.info("BOOTSTRAP: creating hbase:meta region");
     FileSystem fs = rootDir.getFileSystem(conf);
     Path tableDir = CommonFSUtils.getTableDir(rootDir, TableName.META_TABLE_NAME);
-    if (fs.exists(tableDir) && !fs.delete(tableDir, true)) {
+    boolean removeMeta = conf.getBoolean(HConstants.REMOVE_META_ON_RESTART,
```

Comment:
// Checking if meta needs initializing.
    status.setStatus("Initializing meta table if this is a new deploy");
    InitMetaProcedure initMetaProc = null;
    // Print out state of hbase:meta on startup; helps debugging.
    RegionState rs = this.assignmentManager.getRegionStates().
        getRegionState(RegionInfoBuilder.FIRST_META_REGIONINFO);
    LOG.info("hbase:meta {}", rs);
    if (rs != null && rs.isOffline()) {
      Optional<InitMetaProcedure> optProc = procedureExecutor.getProcedures().stream()
        .filter(p -> p instanceof InitMetaProcedure).map(o -> (InitMetaProcedure) o).findAny();
      initMetaProc = optProc.orElseGet(() -> {
        // schedule an init meta procedure if meta has not been deployed yet
        InitMetaProcedure temp = new InitMetaProcedure();
        procedureExecutor.submitProcedure(temp);
        return temp;
      });
    }
So the checks we do see that META location is not there in zk and so it thinks its new deploy.  So here is what we need to tackle.
In cloud redeploy case we will see a pattern where we will have a clusterId in the FS and not in zk. This can be used as an indicator? IMO we should find it (using this way or other) that its a redeploy on an existing datset and all these places, we need to consider that also to decide we need such bootstrap steps.
We should not be doing that with a config way IMO.  Because then in cloud based deploy, what if the 1st time start fail and there is a need for this bootstrap cleaning of META FS dir?
Even the other unknown server case also.  Lets identify clearly this redeploy case and act then only.
Can we pls ave that discuss and conclude on a solution for that and then move forward?

Reply:
> In cloud redeploy case we will see a pattern where we will have a clusterId in the FS and not in zk. This can be used as an indicator? 

I will come back on this later tomorrow, but I agreed with you that we should check explicitly how we define *partial bootstrap* and that `partial meta` need some cleanup.  

also, do you mean if the clusterID did't write to ZK, is it `partial` during bootstrap ?

Reply:
>also, do you mean if the clusterID did't write to ZK, is it partial during bootstrap ?
I am not sure whether that can be really used. I need to check the code.  We need a way to identify the fact that its a cluster redeploy.  Not use some config to identify that.. The HBase system should be smart enough.  So I was just wondering whether this we can use to know that.  May be not.. Need to see.  So my thinking is this that we will make the feature of recreate a cluster on top of existing data a 1st class feature for HBase itself.

Reply:
> > also, do you mean if the clusterID did't write to ZK, is it partial during bootstrap ?
> 
> I am not sure whether that can be really used. I need to check the code. We need a way to identify the fact that its a cluster redeploy. Not use some config to identify that.. The HBase system should be smart enough. So I was just wondering whether this we can use to know that. May be not.. Need to see. So my thinking is this that we will make the feature of recreate a cluster on top of existing data a 1st class feature for HBase itself.

That would be great, let's find a good way to differentiate this case.

Reply:
@anoopsjohn any comments ?

Reply:
Sorry for the delay. Need to read through the analysis what you put above.. What you mention about when, after recreate cluster, the start will hang because of META not getting assigned, is correct.    Can u pls create a sub issue for this case? ie. knowing whether we are starting HM after a recreate (create cluster over existing data)

Reply:
In case of HM start and the bootstrap we create the ClusterID and write to FS and then to zk and then create the META table FS layout.   So in a cluster recreate, we will see clusterID is there in FS and also the META  FS layout but no clusterID in zk. Ya seems we can use this as indication for cluster recreate over existing data.   In HM start, this is some thing we need to check at 1st itself and track.  If this mode is true, later when (if) we do INIT_META_WRITE_FS_LAYOUT , we should not delete the META dir.   As part of the Bootstrap when we write that proc to MasterProcWal, we can include this mode (boolean) info also.  This is a protobuf message anyways.  So even if this HM got killed and restarted (at a point where the clusterId was written to zk but the Meta FS layout part was not reached) we can use the info added as part of the bootstrap wal entry and make sure NOT to delete the meta dir.
Can we do this part alone in a sub task and a provide a patch pls?  This is very key part.. That is why better we can fine tune this with all diff testcases. Sounds good?

Reply:
sounds right to me, as you suggested we put this PR on-held and depends on the new sub-task. I will try to send another JIRA and PR out in a few days and refer to the conversation we discussed here. 

Thanks again Anoop

Reply:
> we should not delete the META dir.

Sorry for harping on an implementation detail: let's sideline meta and not delete please :).

> Can we do this part alone in a sub task and a provide a patch pls? This is very key part..

This seems like a very reasonable starting point. Like Anoop points out, if we can be very sure that we will only trigger this case when we are absolutely sure we're in the "cloud recreate" situation, that will bring a lot of confidence.

> I will try to send another JIRA and PR out in a few days and refer to the conversation we discussed here.

Lazy Josh: did you get a new Jira created already for this?

Reply:
@joshelser the new JIRA is [HBASE-24833](https://issues.apache.org/jira/browse/HBASE-24833) and the discussion are mainly on the new [PR#2237](https://github.com/apache/hbase/pull/2237). I may need to send email to dev@ list for a boarder discussion if we should not depend on the data on zookeeper (that will help us to prevent deleting the meta directory)

Reply:
@josthe new JIRA is [HBASE-24833](https://issues.apache.org/jira/browse/HBASE-24833) and the discussion are mainly on the new [PR#2237](https://github.com/apache/hbase/pull/2237). I may need to send email to dev@ list for a boarder discussion if we should not depend on the data on zookeeper (that will help us to prevent deleting the meta directory)